File: windows/star_term/macros_panel.py
from PySide6.QtCore import Signal
from PySide6.QtWidgets import (
    QCheckBox, QDialog, QDialogButtonBox, QFormLayout, QHBoxLayout, QLabel,
    QLineEdit, QListWidget, QListWidgetItem, QPlainTextEdit, QPushButton,
    QVBoxLayout, QWidget,
)

from .config import load_macros, save_macros


class MacroEditDialog(QDialog):
    def __init__(self, parent=None, macro=None):
        super().__init__(parent)
        self.setWindowTitle("Macro")
        macro = macro or {}

        self.name_edit = QLineEdit(macro.get("name", ""))

        self.commands_edit = QPlainTextEdit(macro.get("commands", ""))
        self.commands_edit.setPlaceholderText(
            "Commands/keystrokes to send, one per line.\n"
            "Each line but the last is sent followed by Enter."
        )
        self.commands_edit.setMinimumSize(360, 160)

        self.auto_execute_checkbox = QCheckBox(
            "Press Enter after the last line (execute immediately)"
        )
        self.auto_execute_checkbox.setChecked(macro.get("auto_execute", False))

        form = QFormLayout()
        form.addRow("Name:", self.name_edit)
        form.addRow("Commands:", self.commands_edit)
        form.addRow(self.auto_execute_checkbox)

        buttons = QDialogButtonBox(QDialogButtonBox.Ok | QDialogButtonBox.Cancel)
        buttons.accepted.connect(self.accept)
        buttons.rejected.connect(self.reject)

        layout = QVBoxLayout(self)
        layout.addLayout(form)
        layout.addWidget(buttons)

    def get_macro(self) -> dict:
        return {
            "name": self.name_edit.text().strip() or "Macro",
            "commands": self.commands_edit.toPlainText(),
            "auto_execute": self.auto_execute_checkbox.isChecked(),
        }


class MacrosPanel(QWidget):
    """Saved keystroke/command sequences that can be sent ("pasted") to the
    active remote session, or broadcast to all sessions in Multi-Exec View."""

    run_requested = Signal(str, bool)

    def __init__(self, parent=None):
        super().__init__(parent)
        self.macros = load_macros()

        self.list_widget = QListWidget()
        self.list_widget.itemDoubleClicked.connect(self._on_run)

        add_btn = QPushButton("Add")
        add_btn.clicked.connect(self._on_add)
        edit_btn = QPushButton("Edit")
        edit_btn.clicked.connect(self._on_edit)
        remove_btn = QPushButton("Remove")
        remove_btn.clicked.connect(self._on_remove)

        edit_row = QWidget()
        edit_layout = QHBoxLayout(edit_row)
        edit_layout.setContentsMargins(0, 0, 0, 0)
        edit_layout.setSpacing(6)
        edit_layout.addWidget(add_btn)
        edit_layout.addWidget(edit_btn)
        edit_layout.addWidget(remove_btn)

        run_btn = QPushButton("Send to Session")
        run_btn.clicked.connect(self._on_run)

        note = QLabel(
            "Sends the macro's commands to the active session, or to all "
            "sessions when Multi-Exec View is on. Unless 'Execute "
            "immediately' is set, the last line is typed but not run."
        )
        note.setObjectName("mutedNote")
        note.setWordWrap(True)

        title = QLabel("Macros")
        title.setObjectName("sectionTitle")

        layout = QVBoxLayout(self)
        layout.setContentsMargins(8, 8, 8, 8)
        layout.setSpacing(8)
        layout.addWidget(title)
        layout.addWidget(self.list_widget)
        layout.addWidget(edit_row)
        layout.addWidget(run_btn)
        layout.addWidget(note)

        self._populate()

    def _populate(self):
        self.list_widget.clear()
        for macro in self.macros:
            self.list_widget.addItem(QListWidgetItem(macro["name"]))

    def _current_index(self):
        row = self.list_widget.currentRow()
        return row if 0 <= row < len(self.macros) else None

    def _on_add(self):
        dialog = MacroEditDialog(self)
        if dialog.exec():
            self.macros.append(dialog.get_macro())
            save_macros(self.macros)
            self._populate()

    def _on_edit(self):
        idx = self._current_index()
        if idx is None:
            return
        dialog = MacroEditDialog(self, self.macros[idx])
        if dialog.exec():
            self.macros[idx] = dialog.get_macro()
            save_macros(self.macros)
            self._populate()

    def _on_remove(self):
        idx = self._current_index()
        if idx is None:
            return
        del self.macros[idx]
        save_macros(self.macros)
        self._populate()

    def _on_run(self):
        idx = self._current_index()
        if idx is None:
            return
        macro = self.macros[idx]
        self.run_requested.emit(macro["commands"], macro.get("auto_execute", False))

File: windows-cpp/src/macrospanel.cpp
#include "macrospanel.h"
#include "config.h"

#include <QCheckBox>
#include <QDialog>
#include <QDialogButtonBox>
#include <QFormLayout>
#include <QHBoxLayout>
#include <QJsonObject>
#include <QLabel>
#include <QLineEdit>
#include <QListWidget>
#include <QPlainTextEdit>
#include <QPushButton>
#include <QVBoxLayout>

// -----------------------------------------------------------------------
// MacroEditDialog (inner)
// -----------------------------------------------------------------------
class MacroEditDialog : public QDialog {
public:
    MacroEditDialog(QWidget *parent = nullptr, const QJsonObject &macro = {})
        : QDialog(parent)
    {
        setWindowTitle("Macro");

        m_nameEdit = new QLineEdit(macro.value("name").toString());
        m_commandsEdit = new QPlainTextEdit(macro.value("commands").toString());
        m_commandsEdit->setPlaceholderText(
            "Commands/keystrokes to send, one per line.\n"
            "Each line but the last is sent followed by Enter.");
        m_commandsEdit->setMinimumSize(360, 160);

        m_autoExecCheck = new QCheckBox(
            "Press Enter after the last line (execute immediately)");
        m_autoExecCheck->setChecked(macro.value("auto_execute").toBool(false));

        QFormLayout *form = new QFormLayout;
        form->addRow("Name:", m_nameEdit);
        form->addRow("Commands:", m_commandsEdit);
        form->addRow(m_autoExecCheck);

        QDialogButtonBox *buttons = new QDialogButtonBox(
            QDialogButtonBox::Ok | QDialogButtonBox::Cancel);
        connect(buttons, &QDialogButtonBox::accepted, this, &QDialog::accept);
        connect(buttons, &QDialogButtonBox::rejected, this, &QDialog::reject);

        QVBoxLayout *layout = new QVBoxLayout(this);
        layout->addLayout(form);
        layout->addWidget(buttons);
    }

    QJsonObject getMacro() const {
        QJsonObject m;
        QString name = m_nameEdit->text().trimmed();
        m["name"]         = name.isEmpty() ? "Macro" : name;
        m["commands"]     = m_commandsEdit->toPlainText();
        m["auto_execute"] = m_autoExecCheck->isChecked();
        return m;
    }

private:
    QLineEdit     *m_nameEdit      = nullptr;
    QPlainTextEdit *m_commandsEdit = nullptr;
    QCheckBox     *m_autoExecCheck = nullptr;
};

// -----------------------------------------------------------------------
// MacrosPanel
// -----------------------------------------------------------------------
MacrosPanel::MacrosPanel(QWidget *parent) : QWidget(parent) {
    m_macros = loadMacros();

    QLabel *title = new QLabel("Macros");
    title->setObjectName("sectionTitle");

    m_listWidget = new QListWidget;
    connect(m_listWidget, &QListWidget::itemDoubleClicked,
            this, &MacrosPanel::onRun);

    QPushButton *addBtn    = new QPushButton("Add");
    QPushButton *editBtn   = new QPushButton("Edit");
    QPushButton *removeBtn = new QPushButton("Remove");
    connect(addBtn,    &QPushButton::clicked, this, &MacrosPanel::onAdd);
    connect(editBtn,   &QPushButton::clicked, this, &MacrosPanel::onEdit);
    connect(removeBtn, &QPushButton::clicked, this, &MacrosPanel::onRemove);

    QWidget *editRow = new QWidget;
    QHBoxLayout *el = new QHBoxLayout(editRow);
    el->setContentsMargins(0,0,0,0);
    el->setSpacing(6);
    el->addWidget(addBtn);
    el->addWidget(editBtn);
    el->addWidget(removeBtn);

    QPushButton *runBtn = new QPushButton("Send to Session");
    connect(runBtn, &QPushButton::clicked, this, &MacrosPanel::onRun);

    QLabel *note = new QLabel(
        "Sends the macro's commands to the active session, or to all "
        "sessions when Multi-Exec View is on. Unless 'Execute "
        "immediately' is set, the last line is typed but not run.");
    note->setObjectName("mutedNote");
    note->setWordWrap(true);

    QVBoxLayout *layout = new QVBoxLayout(this);
    layout->setContentsMargins(8,8,8,8);
    layout->setSpacing(8);
    layout->addWidget(title);
    layout->addWidget(m_listWidget, 1);
    layout->addWidget(editRow);
    layout->addWidget(runBtn);
    layout->addWidget(note);

    populate();
}

void MacrosPanel::populate() {
    m_listWidget->clear();
    for (auto v : m_macros)
        m_listWidget->addItem(v.toObject().value("name").toString());
}

int MacrosPanel::currentIndex() const {
    int row = m_listWidget->currentRow();
    return (row >= 0 && row < m_macros.size()) ? row : -1;
}

void MacrosPanel::onAdd() {
    MacroEditDialog dlg(this);
    if (dlg.exec()) {
        m_macros.append(dlg.getMacro());
        saveMacros(m_macros);
        populate();
    }
}

void MacrosPanel::onEdit() {
    int idx = currentIndex();
    if (idx < 0) return;
    MacroEditDialog dlg(this, m_macros[idx].toObject());
    if (dlg.exec()) {
        m_macros[idx] = dlg.getMacro();
        saveMacros(m_macros);
        populate();
    }
}

void MacrosPanel::onRemove() {
    int idx = currentIndex();
    if (idx < 0) return;
    m_macros.removeAt(idx);
    saveMacros(m_macros);
    populate();
}

void MacrosPanel::onRun() {
    int idx = currentIndex();
    if (idx < 0) return;
    QJsonObject macro = m_macros[idx].toObject();
    emit runRequested(macro.value("commands").toString(),
                      macro.value("auto_execute").toBool(false));
}
